File: js/src/builtins/funcContains.js
import { JPLTypeError } from '../library';

function builtin(runtime, signal, next, input, arg0) {
  const value = runtime.unwrapValue(input);
  const t = runtime.type(value);

  const token = runtime.unwrapValue(arg0 ?? null);
  const tt = runtime.type(token);

  switch (t) {
    case 'array':
      return next(value.some((item) => runtime.equals(item, token)));

    case 'string':
      if (tt === 'string') return next(value.includes(token));
      break;

    case 'object':
      return next(Object.values(value).some((item) => runtime.equals(item, token)));

    default:
  }

  throw new JPLTypeError('%s (%*<100v) cannot contain %s (%*<100v)', t, value, tt, token);
}

export default builtin;

File: go/builtins/funcContains.go
package builtins

import (
	"strings"

	"github.com/2manyvcos/jpl/go/jpl"
	"github.com/2manyvcos/jpl/go/library"
)

var funcContains jpl.JPLFunc = func(runtime jpl.JPLRuntime, signal jpl.JPLRuntimeSignal, next jpl.JPLPiper, input any, args ...any) ([]any, error) {
	value, err := library.UnwrapValue(input)
	if err != nil {
		return nil, err
	}
	t, err := library.Type(value)
	if err != nil {
		return nil, err
	}

	var token any
	if len(args) > 0 {
		token, err = library.UnwrapValue(args[0])
		if err != nil {
			return nil, err
		}
	}
	tt, err := library.Type(token)
	if err != nil {
		return nil, err
	}

	switch t {
	case jpl.JPLT_ARRAY:
		for _, item := range value.([]any) {
			equals, err := library.Equals(item, token)
			if err != nil {
				return nil, err
			}
			if equals {
				return next.Pipe(true)
			}
		}
		return next.Pipe(false)

	case jpl.JPLT_STRING:
		if tt == jpl.JPLT_STRING {
			return next.Pipe(strings.Contains(value.(string), token.(string)))
		}

	case jpl.JPLT_OBJECT:
		for _, item := range value.(map[string]any) {
			equals, err := library.Equals(item, token)
			if err != nil {
				return nil, err
			}
			if equals {
				return next.Pipe(true)
			}
		}
		return next.Pipe(false)

	default:
	}

	return nil, library.ThrowAny(library.NewTypeError("%s (%*<100v) cannot contain %s (%*<100v)", string(t), value, string(tt), token))
}
